### PR TITLE
Block d

### PR DESCRIPTION
**Pull request by Ozymandy**

### CODE REVIEW

File: src/main/java/org/tc/controllers/CourseController.java [hunk 1]
```diff
@@ -64,16 +82,18 @@ public ModelAndView details(@PathVariable("id") int id) {
...
         mav.addObject(HEADER_TITLE, "Course details");
-        mav.addObject("course", courseDetailsDTOConverter.convert(course));
+        mav.addObject(ONE_COURSE_OBJECT_NAME, courseDetailsDTOConverter.convert(course));
         return mav;
     }
 
     @Secured("Lecturer")
     @RequestMapping(value = {"/courses/create"}, method = RequestMethod.GET)
     public ModelAndView create() {
         ModelAndView mav = new ModelAndView("classpath:views/create");
```
Comment (UladzimirKapylou) [inline src/main/java/org/tc/controllers/CourseController.java]: Why you have not refactored this string literal into constant?

File: src/main/java/org/tc/controllers/CourseController.java [hunk 2]
```diff
@@ -111,17 +136,18 @@ public ModelAndView updateGet(@PathVariable("id") int id) {
             method = RequestMethod.POST)
     public ModelAndView updatePost(@PathVariable("id") int id,
                                    @Valid
-                                   @ModelAttribute("course") Course course,
+                                   @ModelAttribute("course") CourseForm course,
                                    BindingResult bindingResult) {
+        ModelAndView mav = new ModelAndView("classpath:views/" + UPDATE_VIEW_NAME);
```
Comment (UladzimirKapylou) [inline src/main/java/org/tc/controllers/CourseController.java]: 1. You use "classpath:views/" in dozens of places, but it is still not a constant.
2. Concatenation here looks not very pretty

File: src/main/java/org/tc/controllers/CourseController.java [hunk 3]
```diff
@@ -208,8 +234,24 @@ public ModelAndView participants(@PathVariable("courseId") int id) {
...
     }
+
+    @RequestMapping(value = {"/mycourses"}, method = RequestMethod.GET)
+    public ModelAndView myCourses(@ModelAttribute("category") Category category) {
+        ModelAndView mav = new ModelAndView("classpath:views/mycourses");
+        mav.addObject(HEADER_TITLE, "My courses");
+        List<Course> myCourses = userService.getMyCourseList();
+        if (category.getCategoryName() != null &&
+                !category.getCategoryName().equals("All")) {
+            myCourses = myCourses.stream().filter(course -> course.getCategory()
```
Comment (UladzimirKapylou) [inline src/main/java/org/tc/controllers/CourseController.java]: Again, bad place to do this. The maximum part of application logic should be done in services

File: src/main/java/org/tc/dao/category/CategoryDao.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package org.tc.dao.category;
+
+import org.tc.models.Category;
+
+import java.util.List;
+
+public interface CategoryDao {
+    void create(Category newCategory);
```
Comment (UladzimirKapylou) [inline src/main/java/org/tc/dao/category/CategoryDao.java]: where do you use this method?

File: src/main/java/org/tc/dao/category/CategoryDao.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+    List<Category> getAll();
+
+    Category getCategoryByName(String categoryName);
+
+    Category getByCategoryId(int id);
+
+    void delete(Category category);
```
Comment (UladzimirKapylou) [inline src/main/java/org/tc/dao/category/CategoryDao.java]: and this?

File: src/main/java/org/tc/models/User.java [hunk 6]
```diff
@@ -35,12 +35,13 @@
     private Role role;
     @OneToMany
     @JoinColumn(name = "userid")
+    private List<Course> ownerCourses;
```
Comment (UladzimirKapylou) [inline src/main/java/org/tc/models/User.java]: probably, ownedCourses

File: src/main/java/org/tc/models/User.java [hunk 7]
```diff
@@ -35,12 +35,13 @@
...
+    @OneToMany
+    @JoinColumn(name = "userid")
```
Comment (UladzimirKapylou) [inline src/main/java/org/tc/models/User.java]: userid is almost senseless here. Why don't you use such names as ownerId, subscriberId or participantId?

File: src/main/java/org/tc/models/forms/CourseForm.java [hunk 8]
```diff
@@ -17,7 +17,26 @@
     private String description;
     @NotEmpty(message = "Blank.Course.Link")
     private String links;
+    @NotEmpty(message = "Blank.Course.Category")
+    private String category;
     private String user;
+    private int id;
```
Comment (UladzimirKapylou) [inline src/main/java/org/tc/models/forms/CourseForm.java:23]: Is this courseFormId?

File: src/main/java/org/tc/utils/converters/CourseConverter.java [hunk 9]
```diff
@@ -5,21 +5,36 @@
...
         course.setName(source.getName());
         course.setDescription(source.getDescription());
         course.setLinks(source.getLinks());
+        course.setId(source.getId());
+        course.setCategory(categoryService
+                .getCategoryByName(source.getCategory()));
         course.setOwner(userService.getByName(source.getUser()));
         return course;
     }
+    public CourseForm reverseConvert(Course course){
```
Comment (UladzimirKapylou) [inline src/main/java/org/tc/utils/converters/CourseConverter.java]: do you really need 'reverse' word here?

File: src/main/resources/views/mycourses.html [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    <link rel="stylesheet" type="text/css"
+          th:href="@{/resources/css/style.css}"/>
+</head>
+<body>
+<div class="container">
+    <header th:replace="header :: header"></header>
+    <div class="courses-top-control">
+        <a class="btn" href="/courses">Courses</a>
+        <div class="search-box">
+            <form class="form-search" th:action="@{/mycourses}" th:object="${category}" method="GET">
```
Comment (UladzimirKapylou) [inline src/main/resources/views/mycourses.html:14]: do you really need separate template for this?

**Review by UladzimirKapylou — CHANGES_REQUESTED** [review]
You did exactly the same mistakes as I found during previous code review. You don't grow. Please, be more attentive.

File: src/main/java/org/tc/controllers/CourseController.java (hunk 2 above)
Comment (Ozymandy) [reply src/main/java/org/tc/controllers/CourseController.java]: Ok, do I need to make constant all view paths?

File: src/main/java/org/tc/controllers/CourseController.java (hunk 1 above)
Comment (Ozymandy) [reply src/main/java/org/tc/controllers/CourseController.java]: Which one? mav.addObject("course")? Or classpath:views/create?

File: src/main/java/org/tc/models/forms/CourseForm.java (hunk 8 above)
Comment (Ozymandy) [reply src/main/java/org/tc/models/forms/CourseForm.java:23]: It's courseId, but I need to represent it in form

File: src/main/java/org/tc/utils/converters/CourseConverter.java (hunk 9 above)
Comment (Ozymandy) [reply src/main/java/org/tc/utils/converters/CourseConverter.java]: Yes, reverse convert. This method converts course to courseForm.This is reverse of convert()

File: src/main/resources/views/mycourses.html (hunk 10 above)
Comment (Ozymandy) [reply src/main/resources/views/mycourses.html:14]: Yes, there are one or two differences in buttons between courses and mycourses

**Pull request merged**